src/glog/logging.h.in

```diff
@@ -1116,44 +1116,38 @@ namespace google {
 
 #if @ac_cv_cxx11_atomic@ && __cplusplus >= 201103L
 #define SOME_KIND_OF_LOG_EVERY_N(severity, n, what_to_do) \
-  static std::atomic<int> LOG_OCCURRENCES(0), LOG_OCCURRENCES_MOD_N(0); \
+  static std::atomic<int> LOG_OCCURRENCES(1); \
   GLOG_IFDEF_THREAD_SANITIZER(AnnotateBenignRaceSized(__FILE__, __LINE__, &LOG_OCCURRENCES, sizeof(int), "")); \
-  GLOG_IFDEF_THREAD_SANITIZER(AnnotateBenignRaceSized(__FILE__, __LINE__, &LOG_OCCURRENCES_MOD_N, sizeof(int), "")); \
-  ++LOG_OCCURRENCES; \
-  if (++LOG_OCCURRENCES_MOD_N > n) LOG_OCCURRENCES_MOD_N -= n; \
-  if (LOG_OCCURRENCES_MOD_N == 1) \
+  int LOG_OCCURRENCES_MOD_N = LOG_OCCURRENCES.fetch_add(1, std::memory_order_relaxed) % n; \
```

Comment:
It is not clear to me why changing to `fetch_add` provides a solution here. `op++` also calls `fetch_add` albeit with different memory ordering. Consequently, you cannot avoid the problem you are trying to solve. To be able to, you would need to perform all the involved operations atomically by protecting concurrent access in a critical section.

Reply:
The problem was  "a series of atomic operations of `LOG_OCCURRENCES_MOD_N`,  but in whole, they are not atomic, which unexpected". Here only the `LOG_OCCURRENCES` is static, the only place we need to protect. Specifically, we only do one atomic operation `fetch_add` on `LOG_OCCURRENCES`.

Reply:
My question is: in what way do the changes fix the problem? The operations are still not atomic as a whole.

Reply:
> The operations are still not atomic as a whole.

Yes. But they don't need to be. There is only one operation that **needs to be atomic**, access to the static variable `LOG_OCCURRENCES`.

> in what way do the changes fix the problem?

For every use of this macro (under different threads), a unique`LOG_OCCURRENCES` value is assigned, incremented by 1. We use a **local** variable `LOG_OCCURRENCES_MOD_N` to get the reminder of `% n`. By now, there is no need for protection or being atmoic. It can be written like
```cpp
int LOG_OCCURRENCES_MOD_N = LOG_OCCURRENCES.fetch_add(1, std::memory_order_relaxed);
if (LOG_OCCURRENCES_MOD_N % n == 1)
```
They are the same.

Reply:
You cannot do that. Once `LOG_OCCURRENCES` overflows you will hit undefined behavior. 

Also, the change from `op++` to `fetch_add` is not clear to me. Why is this necessary?

Reply:
The main idea is to promise every call can have the exact value in the `if` condition statements to represent if it should be logged. When the static atomic variable is `fetch_add` to a local variable, the state is recorded. You'll never get the same value.

Reply:
> Once LOG_OCCURRENCES overflows you will hit undefined behavior.

You're right. This may need more consideration.

> the change from `op++` to `fetch_add` is not clear to me. Why is this necessary? 

`op++` can only be assigned to itself

```cpp
int val = op++; // not atomic, there are two operations. self increment and assign.
int val = op.fetch_add(1) // atomic
```

Reply:
That's not correct. With post increment you will get the previous value.

Reply:
Sorry, It's my bad. You're right. I agree there is no need to use `fetch_add`. `++` is enough.

> Once LOG_OCCURRENCES overflows you will hit undefined behavior.

Use `atomic_uint` instead. Unsigned integer overflow is not a UB. Will this help?